File: androidDemo/src/main/java/com/novage/p2pml/demo/ui/screens/player/PlayerViewModel.kt
package com.novage.p2pml.demo.ui.screens.player

import android.app.Application
import android.content.Context
import android.os.Looper
import androidx.annotation.OptIn
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.viewModelScope
import androidx.media3.common.MediaItem
import androidx.media3.common.Player
import androidx.media3.common.Tracks
import androidx.media3.common.util.Log
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.DefaultLoadControl
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.LoadControl
import androidx.navigation.toRoute
import com.novage.p2pml.P2PMediaLoader
import com.novage.p2pml.P2PMediaLoaderErrorType
import com.novage.p2pml.api.interfaces.Cancellable
import com.novage.p2pml.api.models.CoreConfig
import com.novage.p2pml.api.models.DynamicCoreConfig
import com.novage.p2pml.demo.ui.screens.player.models.MediaTrack
import com.novage.p2pml.demo.ui.screens.player.utils.applyTrackSelection
import com.novage.p2pml.demo.ui.screens.player.utils.getAvailableTracks
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import com.novage.p2pml.demo.ui.navigation.Player as PlayerRoute

private const val HIGH_DEMAND_WINDOW_SEC = 20
private const val PLAYER_MAX_BUFFER_MS = HIGH_DEMAND_WINDOW_SEC * 1000
private const val PLAYER_MIN_BUFFER_MS = 10_000
private const val BUFFER_FOR_PLAYBACK_MS = 2_500
private const val BUFFER_FOR_REBUFFER_MS = 5_000

class PlayerViewModel(application: Application, savedStateHandle: SavedStateHandle) : AndroidViewModel(application) {

    private val _uiState = MutableStateFlow(PlayerUiState())
    val uiState: StateFlow<PlayerUiState> = _uiState.asStateFlow()

    private var currentTracks: Tracks? = null

    private var shouldAutoPlay = true
    var player: ExoPlayer? = null
        private set

    private var p2pLoader: P2PMediaLoader? = null
    private val eventSubscriptions = mutableListOf<Cancellable>()
    private var playerInitializationJob: Job? = null

    init {
        val args = savedStateHandle.toRoute<PlayerRoute>()
        initializePlayer(args.videoUrl, args.customEngineUrl)
    }

    @OptIn(UnstableApi::class)
    fun initializePlayer(manifestUrl: String, customEngineUrl: String?) {
        if (player != null || playerInitializationJob?.isActive == true) return

        playerInitializationJob = viewModelScope.launch {
            val context = getApplication<Application>()
            val exoPlayer = withContext(Dispatchers.IO) {
                ExoPlayer.Builder(context)
                    .setLoadControl(
                        configureBufferSettings()
                    )
                    .setLooper(Looper.getMainLooper())
                    .build()
            }

            ensureActive()

            exoPlayer.addListener(object : Player.Listener {
                override fun onTracksChanged(tracks: Tracks) {
                    currentTracks = tracks

                    val tracks = getAvailableTracks(tracks, exoPlayer.trackSelectionParameters)
                    _uiState.update { it.copy(availableTracks = tracks) }
                }

                override fun onPlaybackStateChanged(playbackState: Int) {
                    if (playbackState != Player.STATE_READY) return

                    _uiState.update { it.copy(isVideoReady = true) }
                }
            })

            exoPlayer.playWhenReady = shouldAutoPlay
            Log.d("PlayerViewModel", "Starting ExoPlayer with P2P Media Loader shouldAutoPlay=$shouldAutoPlay")
            player = exoPlayer

            initializeP2PLoader(context, exoPlayer, manifestUrl, customEngineUrl)
        }
    }

    private fun initializeP2PLoader(
        context: Context,
        exoPlayer: ExoPlayer,
        manifestUrl: String,
        customEngineUrl: String?
    ) {
        P2PMediaLoader.enableLogging()

        val coreConfig = CoreConfig(
            highDemandTimeWindow = HIGH_DEMAND_WINDOW_SEC,
            isP2PDisabled = !shouldAutoPlay,
            simultaneousP2PDownloads = 3,
            webRtcMaxMessageSize = 65535,
            p2pNotReceivingBytesTimeoutMs = 1000,

            validateHTTPSegmentJs = """
            (url, byteRange, data) => {
                console.log(`Validating segment: ${'$'}{url} Range: ${'$'}{byteRange}`);
                return data.byteLength > 0;
            }
        """.trimIndent()
        )

        val loader = P2PMediaLoader(
            context = context,
            coreConfig = coreConfig,
            customEngineUrl = customEngineUrl,
            onReady = {
                val activeLoader = p2pLoader ?: return@P2PMediaLoader
                val p2pUrl = try {
                    activeLoader.getManifestUrl(manifestUrl)
                } catch (_: IllegalStateException) {
                    manifestUrl
                }

                startPlayback(exoPlayer, p2pUrl)
                _uiState.update { it.copy(isP2PActive = true) }
            },
            onError = { type, msg ->
                handleP2PError(type, msg, manifestUrl)
            }
        )

        setupP2PEvents(loader)
        loader.start(exoPlayer)
        p2pLoader = loader
    }

    fun onMessageConsumed() {
        _uiState.update { it.copy(userMessage = null) }
    }

    private fun handleP2PError(type: P2PMediaLoaderErrorType, msg: String, originalUrl: String) {
        val exoPlayer = player ?: return

        when (type) {
            P2PMediaLoaderErrorType.ENGINE_STARTUP_ERROR,
            P2PMediaLoaderErrorType.ENGINE_RUNTIME_ERROR -> {
                startPlayback(exoPlayer, originalUrl)

                _uiState.update {
                    it.copy(
                        isP2PActive = false,
                        userMessage = "P2P Engine failed. Switched to HTTP mode."
                    )
                }
            }

            P2PMediaLoaderErrorType.MANIFEST_LOAD_ERROR,
            P2PMediaLoaderErrorType.MANIFEST_PARSE_ERROR -> {
                _uiState.update {
                    it.copy(
                        fatalError = "Video unavailable: $msg"
                    )
                }
            }

            P2PMediaLoaderErrorType.SEGMENT_DOWNLOAD_ERROR -> TODO()
        }
    }

    fun changeTrack(track: MediaTrack) {
        val player = player ?: return
        val tracks = currentTracks ?: return

        applyTrackSelection(player, track, tracks)

        val updatedTracks = getAvailableTracks(tracks, player.trackSelectionParameters)
        _uiState.update { it.copy(availableTracks = updatedTracks) }
    }

    @OptIn(UnstableApi::class)
    private fun configureBufferSettings(): LoadControl = DefaultLoadControl.Builder()
        .setBufferDurationsMs(
            PLAYER_MIN_BUFFER_MS,
            PLAYER_MAX_BUFFER_MS,
            BUFFER_FOR_PLAYBACK_MS,
            BUFFER_FOR_REBUFFER_MS
        ).build()

    private fun startPlayback(exoPlayer: ExoPlayer, url: String) {
        exoPlayer.setMediaItem(MediaItem.fromUri(url))
        exoPlayer.prepare()
    }

    private fun setupP2PEvents(loader: P2PMediaLoader) {
        eventSubscriptions.add(
            loader.onChunkDownloaded { chunk ->
                _uiState.update { state ->
                    state.copy(
                        totalDownloaded = state.totalDownloaded + chunk.bytesLength,
                        p2pDownloaded = if (chunk.downloadSource == "p2p") {
                            state.p2pDownloaded + chunk.bytesLength
                        } else {
                            state.p2pDownloaded
                        },
                        httpDownloaded = if (chunk.downloadSource == "http") {
                            state.httpDownloaded + chunk.bytesLength
                        } else {
                            state.httpDownloaded
                        }
                    )
                }
            }
        )

        eventSubscriptions.add(
            loader.onChunkUploaded { chunk ->
                _uiState.update { state ->
                    state.copy(uploadTotal = state.uploadTotal + chunk.bytesLength)
                }
            }
        )

        eventSubscriptions.add(
            loader.onPeerConnect { peer ->
                _uiState.update { state ->
                    state.copy(peers = state.peers + peer)
                }
            }
        )

        eventSubscriptions.add(
            loader.onPeerClose { peer ->
                _uiState.update { state ->
                    state.copy(peers = state.peers.filter { it.peerId != peer.peerId })
                }
            }
        )
    }

    override fun onCleared() {
        super.onCleared()
        releaseResources()
    }

    fun play() {
        shouldAutoPlay = true
        player?.play()
        setP2PEnabled(true)
    }

    fun pause() {
        shouldAutoPlay = false
        player?.pause()
        setP2PEnabled(false)
    }

    private fun setP2PEnabled(isEnabled: Boolean) {
        val loader = p2pLoader ?: return

        val config = DynamicCoreConfig(
            isP2PDisabled = !isEnabled
        )

        loader.applyDynamicConfig(config)
    }

    private fun releaseResources() {
        playerInitializationJob?.cancel()

        eventSubscriptions.forEach { it.cancel() }
        eventSubscriptions.clear()

        player?.release()
        player = null

        p2pLoader?.release()
        p2pLoader = null
    }
}

File: iosApp/iosApp/UI/Screens/Player/PlayerViewModel.swift
import Foundation
import AVKit
import Combine
import P2PML

@MainActor
class PlayerViewModel: ObservableObject {
    @Published var uiState = PlayerUiState()
    @Published var player: AVPlayer? = nil

    private var p2pLoader: P2PMediaLoader? = nil
    private var eventSubscriptions: [P2PML.Cancellable] = []
    private var playerItemObserver: Any?
    private var shouldAutoPlay = true

    func initializePlayer(manifestUrl: String, customEngineUrl: String?) {
        guard player == nil else { return }

        P2PMediaLoader.companion.enableLogging()

        // Exact config mapping from Android
        let coreConfig = """
                         {
                             "highDemandTimeWindow": 20,
                             "isP2PDisabled": \(!shouldAutoPlay),
                             "simultaneousP2PDownloads": 3,
                             "webRtcMaxMessageSize": 65535,
                             "p2pNotReceivingBytesTimeoutMs": 1000
                         }
                         """.replacingOccurrences(of: "\n", with: " ")

        let loader = P2PMediaLoader(
            onReady: { [weak self] in
                guard let self = self else { return }
                let p2pUrl = self.p2pLoader?.getManifestUrl(manifestUrl: manifestUrl) ?? manifestUrl
                self.startPlayback(url: p2pUrl)
                self.uiState.isP2PActive = true
            },
            onError: { [weak self] errorType, errorMessage in
                self?.handleP2PError(errorMessage: errorMessage, originalUrl: manifestUrl)
            },
            coreConfig: coreConfig,
            customEngineUrl: customEngineUrl
        )

        setupP2PEvents(loader)

        loader.start(getPlaybackInfo: { [weak self] in
            guard let validPlayer = self?.player else {
                return PlaybackInfo(currentPlayPosition: 0.0, currentPlaybackSpeed: 0.0)
            }
            let currentSeconds = CMTimeGetSeconds(validPlayer.currentTime())
            return PlaybackInfo(
                currentPlayPosition: currentSeconds.isNaN ? 0.0 : currentSeconds,
                currentPlaybackSpeed: Float(validPlayer.rate)
            )
        })

        self.p2pLoader = loader
    }

    private func handleP2PError(errorMessage: String, originalUrl: String) {
        startPlayback(url: originalUrl)
        uiState.isP2PActive = false
        uiState.userMessage = "P2P Engine failed. Switched to HTTP mode."
    }

    private func startPlayback(url: String) {
        guard let urlObj = URL(string: url) else {
            uiState.fatalError = "Invalid URL"
            return
        }

        let playerItem = AVPlayerItem(url: urlObj)
        let newPlayer = AVPlayer(playerItem: playerItem)
        newPlayer.automaticallyWaitsToMinimizeStalling = true // iOS Default LoadControl
        self.player = newPlayer

        // Monitor buffering state
        playerItemObserver = playerItem.observe(\.status, options: [.new]) { [weak self] item, _ in
            DispatchQueue.main.async {
                if item.status == .readyToPlay {
                    self?.uiState.isVideoReady = true
                    // Mock tracks to match Android UI (AVPlayer handles this natively)
                    self?.uiState.availableTracks = AvailableTracks(
                        videoTracks: [MediaTrack(label: "Auto", isSelected: true, isAuto: true)],
                        audioTracks: [MediaTrack(label: "Default", isSelected: true, isAuto: true)]
                    )
                } else if item.status == .failed {
                    self?.uiState.fatalError = "Video unavailable: \(item.error?.localizedDescription ?? "Unknown error")"
                }
            }
        }

        if shouldAutoPlay { newPlayer.play() }
    }

    private func setupP2PEvents(_ loader: P2PMediaLoader) {
        eventSubscriptions.append(loader.onChunkDownloaded { [weak self] details in
            guard let self = self else { return }
            self.uiState.totalDownloaded += Int64(details.bytesLength)
            if details.downloadSource == "p2p" {
                self.uiState.p2pDownloaded += Int64(details.bytesLength)
            } else {
                self.uiState.httpDownloaded += Int64(details.bytesLength)
            }
        })
        eventSubscriptions.append(loader.onChunkUploaded { [weak self] details in
            self?.uiState.uploadTotal += Int64(details.bytesLength)
        })
        eventSubscriptions.append(loader.onPeerConnect { [weak self] _ in
            self?.uiState.peerCount += 1
        })
        eventSubscriptions.append(loader.onPeerClose { [weak self] _ in
            self?.uiState.peerCount = max(0, (self?.uiState.peerCount ?? 1) - 1)
        })
    }

    func onMessageConsumed() { uiState.userMessage = nil }

    func changeTrack(_ track: MediaTrack) {
        // AVPlayer handles HLS variants automatically.
        // Deep integration via AVAssetReader is needed to override this manually in iOS.
    }

    func play() {
        shouldAutoPlay = true
        player?.play()
    }

    func pause() {
        shouldAutoPlay = false
        player?.pause()
    }

    func releaseResources() {
        player?.pause()
        player = nil
        eventSubscriptions.forEach { $0.cancel() }
        eventSubscriptions.removeAll()
        p2pLoader?.release()
        p2pLoader = nil
    }
}
